guidance/guidance_los/scripts/los_guidance_action_server.py

```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env python3
+
+import time
+from dataclasses import dataclass
+
+import numpy as np
+import rclpy
+from geometry_msgs.msg import PoseStamped, Vector3Stamped
+from guidance_los.los_guidance_algorithm import ThirdOrderLOSGuidance
+from nav_msgs.msg import Odometry
+from rclpy.action import ActionServer
+from rclpy.action.server import ServerGoalHandle
+from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.node import Node
+from std_msgs.msg import String
+from transforms3d.euler import quat2euler
+from vortex_msgs.action import NavigateWaypoints
+from vortex_msgs.msg import LOSGuidance
+
+
+@dataclass(slots=True)
+class State:
+    """Data class representing the current state of the vehicle."""
+
+    x: float = 0.0
+    y: float = 0.0
+    z: float = 0.0
+    yaw: float = 0.0
+    pitch: float = 0.0
+
+    def as_ndarray(self) -> np.ndarray:
+        """Convert state to numpy array for guidance calculations."""
+        return np.array([self.x, self.y, self.z, self.yaw, self.pitch])
+
+
+class GuidanceActionServer(Node):
+    """ROS2 Action Server node for waypoint navigation using LOS guidance."""
+
+    def __init__(self):
+        super().__init__('guidance_action_server')
+
+        # Initialize filter status flag
+        self.filter_initialized = False
+
+        # Add debug parameter
+        self.declare_parameter('debug_mode', False)
+        self.debug_mode = self.get_parameter('debug_mode').value
+
+        # Create callback groups for concurrent execution
+        self.action_cb_group = ReentrantCallbackGroup()
+        self.timer_cb_group = ReentrantCallbackGroup()
+
+        # Initialize guidance calculator with third-order filtering
+        self.guidance_calculator = ThirdOrderLOSGuidance()
+
+        # Publishers for guidance commands and monitoring
+        self.guidance_cmd_pub = self.create_publisher(
+            LOSGuidance, '/guidance/LOS_commands', 10
+        )
+
+        # Debug publishers (if debug mode is enabled)
+        if self.debug_mode:
+            self.guidance_ref_pub = self.create_publisher(
+                PoseStamped, '/guidance/debug/reference', 10
+            )
+            self.guidance_error_pub = self.create_publisher(
+                Vector3Stamped, '/guidance/debug/errors', 10
+            )
+            self.log_publisher = self.create_publisher(
+                String, '/guidance/debug/logs', 10
+            )
+
+        # Subscriber for vehicle state updates
+        self.create_subscription(
+            Odometry,
+            '/nucleus/odom',
+            self.odom_callback,
+            10,
+            callback_group=self.timer_cb_group,
+        )
```

Comment:
load the topics from a config file, so it is easy to change them without needing to change the code